plugins/sinks/columbus/sink.go

```diff
@@ -33,27 +35,73 @@ func New(c httpClient) plugins.Syncer {
 }
 
 func (s *Sink) Sink(ctx context.Context, configMap map[string]interface{}, in <-chan interface{}) (err error) {
-	if err = utils.BuildConfig(configMap, &s.config); err != nil {
+	var config Config
+	if err = utils.BuildConfig(configMap, &config); err != nil {
 		return plugins.InvalidConfigError{Type: plugins.PluginTypeSink}
 	}
 
 	for data := range in {
-		if err = s.send(data); err != nil {
+		dataMapper := s.getDataMapper(data, config)
+		data = dataMapper(data)
+		if err = s.send(data, config); err != nil {
 			return
 		}
 	}
 
 	return
 }
 
-func (s *Sink) send(data interface{}) (err error) {
+func (s *Sink) getDataMapper(data interface{}, config Config) func(interface{}) interface{} {
+	if s.cachedDataMapper == nil {
+		s.cachedDataMapper = s.buildDataMapper(data, config)
+	}
+
+	return s.cachedDataMapper
+}
+
+func (s *Sink) buildDataMapper(data interface{}, config Config) func(interface{}) interface{} {
+	if config.FieldsMapper == nil {
+		return s.defaultMapper
+	}
+
+	// build new type
+	value := reflect.ValueOf(data)
+	newType := s.buildNewType(value, config.FieldsMapper)
+
+	return func(d interface{}) interface{} {
+		v := reflect.ValueOf(d)
+		newValue := v.Convert(newType)
+		return newValue.Interface()
+	}
+}
+
+func (s *Sink) buildNewType(value reflect.Value, fieldMap map[string]string) reflect.Type {
+	t := value.Type()
+	sf := make([]reflect.StructField, 0)
+	for i := 0; i < t.NumField(); i++ {
+		sf = append(sf, t.Field(i))
+
+		// check if field will be mapped
+		jsonField, ok := fieldMap[t.Field(i).Name]
+		if !ok {
+			continue // skip if field is not mentioned in the mapper
+		}
+		sf[i].Tag = reflect.StructTag(fmt.Sprintf(`json:"%s"`, jsonField))
```

Comment:
Can it be written concisely? maybe like this.
```
if jsonField, ok := fieldMap[t.Field(i).Name]; ok{
  sf[i].Tag = reflect.StructTag(fmt.Sprintf(`json:"%s"`, jsonField))
}

Reply:
oh was in a habit of exiting early, but since the valid condition is only a one liner we can do it your way.